number_theory/sieve.h

```diff
@@ -0,0 +1,97 @@
+#ifndef NUMBER_THEORY_SIEVE_H_
+#define NUMBER_THEORY_SIEVE_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <bit>
+#include <limits>
+#include <stdexcept>
+#include <vector>
+
+#include "number_theory/utility.h"
+
+namespace tql {
+namespace number_theory {
+
+// Sieve of Euler.
+// It finds all prime numbers under a certain limit.
+// It also provides the factorizations of all numbers under the limit.
+//
+// The numbers are of type |T|, but they are converted to type |U| during
+// multiplication to avoid overflow.
+template <typename T, typename U = uint64_t>
+class Sieve {
+  static_assert(std::numeric_limits<T>::is_integer &&
+                    std::numeric_limits<U>::is_integer,
+                "Sieve must use integer types.");
+
+ public:
+  // The type of numbers used by the Sieve.
+  using type = T;
+
+  // Constructs the Sieve in linear time.
+  explicit Sieve(T num_limit)
+      : num_limit_(std::move(num_limit)), min_prime_factor_(num_limit_, T(0)) {
```

Comment:
I think it would be safer to use `num_limit` instead of `num_limit_` to initialize `min_prime_factor_`. The constructor initializer list's initialization order is not the same as we write them here. For example, it is easy to assume that `num_limit_` gets initialized first since it comes before than `min_prime_factor_`. However, the real reason that `num_limit_` gets initialized first is the fact that it is defined above `min_prime_factor_` in class' private field.

In order to avoid any issues, it is better to keep the order the same in initializer list and class member declaration order. In this case, I just prefer relying on `num_limit` variable since we know that it is already there. Oh, but since we are moving it and it happens first, `num_limit` is not available anymore.

So, do we really need to complicate things this much just to use `std::move` there? Maybe it is not that much worth it. It would be way simpler to just use `num_limit` in constructor list initializations. I don't remember using `std::move` in my constructors for FileSystemScanner constructors in chromium. It might be too much optimization at this point.

Still, I would like to hear your opinion about this subject in general. I agree that using `std::move` is good, but I also think that over-optimization is something dangerous that can easily cause the code base to become too complicated than necessary without providing enough compensation.